# ACTIVIDAD 1

#### Importamos las librerias correspondientes

In [108]:
# Librerías de Python para la manipulación de datos
import pandas as pd
import numpy as np

# Para la visualización de datos
import matplotlib.pyplot as plt
from matplotlib import rcParams
import seaborn as sns
import plotly.express as px

## 1.- HEART DISEASE DATASET

### a) Visualizamos y agregamos las cabeceras del dataset

In [88]:
df = pd.read_csv("processed.cleveland.data",names =["age", "sex", "cp","trestbps","chol","fbs", "restecg", "thalach", "exang","oldpeak","slope", "ca", "thal", "num"])
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  
0    3.0  0.0  6.0    0  
1    2.0  3.0  3.0    2  
2    2.0  2.0  7.0    1  
3    3.0  0.0  3.0    0  
4    1.0  0.0  3.0    0

### b) Señalamos las variables numericas

In [39]:
df_1 = df._get_numeric_data()
df_1.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope  num  
0    3.0    0  
1    2.0    2  
2    2.0    1  
3    3.0    0  
4    1.0    0

Si bien tenemos las variables "numericas" con la sentencia antes ejecutada, podemos deducir leyendo el archivo "heart-disease.names" que las variables numericas son: age, trestbps, chol, thalach, oldpeak.

### Señalamos las variables continuas y discretas

##### AGE

In [40]:
df_1['age'].unique()

array([63., 67., 37., 41., 56., 62., 57., 53., 44., 52., 48., 54., 49.,
       64., 58., 60., 50., 66., 43., 40., 69., 59., 42., 55., 61., 65.,
       71., 51., 46., 45., 39., 68., 47., 34., 35., 29., 70., 77., 38.,
       74., 76.])

La variable edad se considera DISCRETA porque los datos de edad no tienen valores decimales

##### TRESTBPS

In [41]:
df_1['trestbps'].unique()

array([145., 160., 120., 130., 140., 172., 150., 110., 132., 117., 135.,
       112., 105., 124., 125., 142., 128., 170., 155., 104., 180., 138.,
       108., 134., 122., 115., 118., 100., 200.,  94., 165., 102., 152.,
       101., 126., 174., 148., 178., 158., 192., 129., 144., 123., 136.,
       146., 106., 156., 154., 114., 164.])

La variable trestbps se considera DISCRETA porque los datos de trestbps no tienen valores decimales

##### CHOL

In [42]:
df_1['chol'].unique()

array([233., 286., 229., 250., 204., 236., 268., 354., 254., 203., 192.,
       294., 256., 263., 199., 168., 239., 275., 266., 211., 283., 284.,
       224., 206., 219., 340., 226., 247., 167., 230., 335., 234., 177.,
       276., 353., 243., 225., 302., 212., 330., 175., 417., 197., 198.,
       290., 253., 172., 273., 213., 305., 216., 304., 188., 282., 185.,
       232., 326., 231., 269., 267., 248., 360., 258., 308., 245., 270.,
       208., 264., 321., 274., 325., 235., 257., 164., 141., 252., 255.,
       201., 222., 260., 182., 303., 265., 309., 307., 249., 186., 341.,
       183., 407., 217., 288., 220., 209., 227., 261., 174., 281., 221.,
       205., 240., 289., 318., 298., 564., 246., 322., 299., 300., 293.,
       277., 214., 207., 223., 160., 394., 184., 315., 409., 244., 195.,
       196., 126., 313., 259., 200., 262., 215., 228., 193., 271., 210.,
       327., 149., 295., 306., 178., 237., 218., 242., 319., 166., 180.,
       311., 278., 342., 169., 187., 157., 176., 24

La variable chol se considera CONTINUA porque los datos de chol tienen valores decimales

##### THALACH

In [43]:
df_1['thalach'].unique()

array([150., 108., 129., 187., 172., 178., 160., 163., 147., 155., 148.,
       153., 142., 173., 162., 174., 168., 139., 171., 144., 132., 158.,
       114., 151., 161., 179., 120., 112., 137., 157., 169., 165., 123.,
       128., 152., 140., 188., 109., 125., 131., 170., 113.,  99., 177.,
       141., 180., 111., 143., 182., 156., 115., 149., 145., 146., 175.,
       186., 185., 159., 130., 190., 136.,  97., 127., 154., 133., 126.,
       202., 103., 166., 164., 184., 124., 122.,  96., 138.,  88., 105.,
       194., 195., 106., 167.,  95., 192., 117., 121., 116.,  71., 118.,
       181., 134.,  90.])

La variable thalach se considera DISCRETA porque los datos de thalach no tienen valores decimales

##### OLDPEAK

In [44]:
df_1['oldpeak'].unique()

array([2.3, 1.5, 2.6, 3.5, 1.4, 0.8, 3.6, 0.6, 3.1, 0.4, 1.3, 0. , 0.5,
       1.6, 1. , 1.2, 0.2, 1.8, 3.2, 2.4, 2. , 2.5, 2.2, 2.8, 3. , 3.4,
       6.2, 4. , 5.6, 2.9, 0.1, 2.1, 1.9, 4.2, 0.9, 1.1, 3.8, 0.7, 0.3,
       4.4])

La variable oldpeak se considera CONTINUA porque los datos de oldpeak tienen valores decimales

### c) Señalamos las variables categoricas

In [50]:
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  
0    3.0  0.0  6.0    0  
1    2.0  3.0  3.0    2  
2    2.0  2.0  7.0    1  
3    3.0  0.0  3.0    0  
4    1.0  0.0  3.0    0

Aqui tenemos todas las variables del dataset, podemos deducir leyendo el archivo "heart-disease.names" que las variables categoricas son: sex, cp, fbs, exang, slope, thal, num.

### Señalamos las variables ordinales y nominales

##### SEX

In [46]:
df['sex'].unique()

array([1., 0.])

La variable sex es NOMINAL; ya que puede variar el orden de sus valores y el concepto de la variable no cambiaria.

##### CP

In [47]:
df['cp'].unique()

array([1., 4., 3., 2.])

La variable cp es ORDINAL; ya que no puede variar el orden de sus valores y el concepto de la variable cambiaria.

##### FBS

In [48]:
df['fbs'].unique()

array([1., 0.])

La variable fbs es NOMINAL; ya que puede variar el orden de sus valores y el concepto de la variable no cambiaria.

##### EXANG

In [33]:
df['exang'].unique()

array([0., 1.])

La variable exang es NOMINAL; ya que puede variar el orden de sus valores y el concepto de la variable no cambiaria.

##### THAL

In [51]:
df['thal'].unique()

array(['6.0', '3.0', '7.0', '?'], dtype=object)

La variable thal es NOMINAL; ya que puede variar el orden de sus valores y el concepto de la variable no cambiaria.

##### NUM

In [52]:
df['num'].unique()

array([0, 2, 1, 3, 4], dtype=int64)

La variable num es NOMINAL; ya que puede variar el orden de sus valores y el concepto de la variable no cambiaria. Considerar que si es diferente de "cero" se clasifica 

### d) Missing values en el dataset

In [91]:
# reemplazar los valores "?" por valores nulos
df.replace("?", pd.NaT, inplace=True)

# ver las columnas que tienen valores nulos
print(df.isnull().sum())

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64


Revisando el dataset validamos que habian campos en donde se encontraba el simbolo "?" por la cual con la sentencia anterior reemplazamos y lo convertimos en NULL. Validando que si hay valores faltantes, el simbolo
que se utilizo es "?" y en este caso no se utiliza mas de un simbolo.

### e) Variable "thal"

In [92]:
df['thal'].unique()

array(['6.0', '3.0', '7.0', nan], dtype=object)

Aqui obtenemos la lista de los posibles valores que puede tomar la variable "thal"

In [107]:
df.info()
thal_df = pd.DataFrame(df['thal'].unique())
thal_df.head().tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        299 non-null    object 
 12  thal      301 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


0
0  6.0
1  3.0
2  7.0
3  NaN

Utilizando métodos de Python, obtén la lista de valores posibles que
puede tomar la variable “thal”. No se pueden ordenar estos valores; porque son de tipo object.
Se podria considerar ordinales; ya que el orden importa.

## 2.- CENSUS INCOME DATASET

### a) Visualizamos y agregamos las cabeceras del dataset

In [159]:
df_adult = pd.read_csv("adult.data",names =["age", "workclass", "fnlwgt","education","education-num","marital-status", "occupation", "relationship", "race","sex","capital-gain", "capital-loss", "hours-per-week", "native-country",""])
df_adult

age          workclass  fnlwgt    education  education-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

               
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

### b) Señalamos las variables numericas y categoricas

#### Variables numericas

In [117]:
dfa_1 = df_adult._get_numeric_data()
dfa_1.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week
0   39   77516             13          2174             0              40
1   50   83311             13             0             0              13
2   38  215646              9             0             0              40
3   53  234721              7             0             0              40
4   28  338409             13             0             0              40

Podemos observar despues de ejecutar el codigo anterior, que las variables numericas son: age, fnlwgt, education-num, capital-gain, capital-loss y hours-per-week

##### Variables continuas o discretas

#### AGE

In [118]:
dfa_1['age'].unique()

array([39, 50, 38, 53, 28, 37, 49, 52, 31, 42, 30, 23, 32, 40, 34, 25, 43,
       54, 35, 59, 56, 19, 20, 45, 22, 48, 21, 24, 57, 44, 41, 29, 18, 47,
       46, 36, 79, 27, 67, 33, 76, 17, 55, 61, 70, 64, 71, 68, 66, 51, 58,
       26, 60, 90, 75, 65, 77, 62, 63, 80, 72, 74, 69, 73, 81, 78, 88, 82,
       83, 84, 85, 86, 87], dtype=int64)

La variable edad se considera DISCRETA porque los datos de edad no tienen valores decimales

#### FNLWGT

In [119]:
dfa_1['fnlwgt'].unique()

array([ 77516,  83311, 215646, ...,  34066,  84661, 257302], dtype=int64)

La variable fnlwgt se considera CONTINUA porque los datos de fnlwgt puede tener valores decimales

#### EDUCATION-NUM

In [120]:
dfa_1['education-num'].unique()

array([13,  9,  7, 14,  5, 10, 12, 11,  4, 16, 15,  3,  6,  2,  1,  8],
      dtype=int64)

La variable education-num se considera DISCRETA porque los datos de education-num no tienen valores decimales

#### CAPITAL-GAIN

In [121]:
dfa_1['capital-gain'].unique()

array([ 2174,     0, 14084,  5178,  5013,  2407, 14344, 15024,  7688,
       34095,  4064,  4386,  7298,  1409,  3674,  1055,  3464,  2050,
        2176,   594, 20051,  6849,  4101,  1111,  8614,  3411,  2597,
       25236,  4650,  9386,  2463,  3103, 10605,  2964,  3325,  2580,
        3471,  4865, 99999,  6514,  1471,  2329,  2105,  2885, 25124,
       10520,  2202,  2961, 27828,  6767,  2228,  1506, 13550,  2635,
        5556,  4787,  3781,  3137,  3818,  3942,   914,   401,  2829,
        2977,  4934,  2062,  2354,  5455, 15020,  1424,  3273, 22040,
        4416,  3908, 10566,   991,  4931,  1086,  7430,  6497,   114,
        7896,  2346,  3418,  3432,  2907,  1151,  2414,  2290, 15831,
       41310,  4508,  2538,  3456,  6418,  1848,  3887,  5721,  9562,
        1455,  2036,  1831, 11678,  2936,  2993,  7443,  6360,  1797,
        1173,  4687,  6723,  2009,  6097,  2653,  1639, 18481,  7978,
        2387,  5060], dtype=int64)

La variable capital-gain se considera CONTINUA porque los datos de capital-gain puede tener valores decimales

#### CAPITAL-GAIN

In [125]:
dfa_1['capital-gain'].unique()

array([ 2174,     0, 14084,  5178,  5013,  2407, 14344, 15024,  7688,
       34095,  4064,  4386,  7298,  1409,  3674,  1055,  3464,  2050,
        2176,   594, 20051,  6849,  4101,  1111,  8614,  3411,  2597,
       25236,  4650,  9386,  2463,  3103, 10605,  2964,  3325,  2580,
        3471,  4865, 99999,  6514,  1471,  2329,  2105,  2885, 25124,
       10520,  2202,  2961, 27828,  6767,  2228,  1506, 13550,  2635,
        5556,  4787,  3781,  3137,  3818,  3942,   914,   401,  2829,
        2977,  4934,  2062,  2354,  5455, 15020,  1424,  3273, 22040,
        4416,  3908, 10566,   991,  4931,  1086,  7430,  6497,   114,
        7896,  2346,  3418,  3432,  2907,  1151,  2414,  2290, 15831,
       41310,  4508,  2538,  3456,  6418,  1848,  3887,  5721,  9562,
        1455,  2036,  1831, 11678,  2936,  2993,  7443,  6360,  1797,
        1173,  4687,  6723,  2009,  6097,  2653,  1639, 18481,  7978,
        2387,  5060], dtype=int64)

La variable capital-loss se considera CONTINUA porque los datos de capital-loss puede tener valores decimales

#### HOURS-PER-WEEK 

In [124]:
dfa_1['hours-per-week'].unique()

array([40, 13, 16, 45, 50, 80, 30, 35, 60, 20, 52, 44, 15, 25, 38, 43, 55,
       48, 58, 32, 70,  2, 22, 56, 41, 28, 36, 24, 46, 42, 12, 65,  1, 10,
       34, 75, 98, 33, 54,  8,  6, 64, 19, 18, 72,  5,  9, 47, 37, 21, 26,
       14,  4, 59,  7, 99, 53, 39, 62, 57, 78, 90, 66, 11, 49, 84,  3, 17,
       68, 27, 85, 31, 51, 77, 63, 23, 87, 88, 73, 89, 97, 94, 29, 96, 67,
       82, 86, 91, 81, 76, 92, 61, 74, 95], dtype=int64)

La variable hours-per-week se considera CONTINUA porque los datos de hours-per-week puede tener valores decimales

#### Variables categoricas

In [137]:
df_adult.head()

age               32561
workclass         32561
fnlwgt            32561
education         32561
education-num     32561
marital-status    32561
occupation        32561
relationship      32561
race              32561
sex               32561
capital-gain      32561
capital-loss      32561
hours-per-week    32561
native-country    32561
                  32561
dtype: int64

#### WORKCLASS

#### EDUCATION

In [ ]:
#### MARITAL-STATUS

In [ ]:
#### OCCUPATION

In [ ]:
#### RELATIONSHIP

In [ ]:
#### RACE

In [ ]:
#### SEX

In [ ]:
#### NATIVE-COUNTRY

### c) Registros repetidos

In [162]:
df2_sal = df_adult[df_adult.duplicated()]
df2_sal.head()

age workclass  fnlwgt      education  education-num  marital-status  \
4881    25   Private  308144      Bachelors             13   Never-married   
5104    90   Private   52386   Some-college             10   Never-married   
9171    21   Private  250051   Some-college             10   Never-married   
11631   20   Private  107658   Some-college             10   Never-married   
13084   25   Private  195994        1st-4th              2   Never-married   

             occupation    relationship                 race      sex  \
4881       Craft-repair   Not-in-family                White     Male   
5104      Other-service   Not-in-family   Asian-Pac-Islander     Male   
9171     Prof-specialty       Own-child                White   Female   
11631      Tech-support   Not-in-family                White   Female   
13084   Priv-house-serv   Not-in-family                White   Female   

       capital-gain  capital-loss  hours-per-week  native-country          
4881              0             0              40          Mexico   <=50K  
5104              0             0              35   United-States   <=50K  
9171              0             0              10   United-States   <=50K  
11631             0             0              10   United-States   <=50K  
13084             0             0              40       Guatemala   <=50K

Con la sentencia antes ejecutada deducimos que si existen registros duplicados

In [139]:
df_adult.count()

age               32561
workclass         32561
fnlwgt            32561
education         32561
education-num     32561
marital-status    32561
occupation        32561
relationship      32561
race              32561
sex               32561
capital-gain      32561
capital-loss      32561
hours-per-week    32561
native-country    32561
                  32561
dtype: int64

aqui validamos que existen 32561 registros en el dataframe original

In [163]:
df2_sal = df2_sal.groupby(df2_sal.columns.tolist()).apply(lambda x: x.index.tolist()).values.tolist()
#print (df2_sal)

print(df2_sal.__sizeof__())


224


Se tiene que duplicados tenemos 224. Y en total elementos unicos tenemos 32337

### d) Dataframe de los repetidos

In [210]:
df_dup = df_adult[df_adult.duplicated()]
df_dup

age          workclass  fnlwgt      education  education-num  \
4881    25            Private  308144      Bachelors             13   
5104    90            Private   52386   Some-college             10   
9171    21            Private  250051   Some-college             10   
11631   20            Private  107658   Some-college             10   
13084   25            Private  195994        1st-4th              2   
15059   21            Private  243368      Preschool              1   
17040   46            Private  173243        HS-grad              9   
18555   30            Private  144593        HS-grad              9   
18698   19            Private   97261        HS-grad              9   
21318   19            Private  138153   Some-college             10   
21490   19            Private  146679   Some-college             10   
21875   49            Private   31267        7th-8th              4   
22300   25            Private  195994        1st-4th              2   
22367   44            Private  367749      Bachelors             13   
22494   49   Self-emp-not-inc   43479   Some-college             10   
25872   23            Private  240137        5th-6th              3   
26313   28            Private  274679        Masters             14   
28230   27            Private  255582        HS-grad              9   
28522   42            Private  204235   Some-college             10   
28846   39            Private   30916        HS-grad              9   
29157   38            Private  207202        HS-grad              9   
30845   46            Private  133616   Some-college             10   
31993   19            Private  251579   Some-college             10   
32404   35            Private  379959        HS-grad              9   

            marital-status          occupation    relationship  \
4881         Never-married        Craft-repair   Not-in-family   
5104         Never-married       Other-service   Not-in-family   
9171         Never-married      Prof-specialty       Own-child   
11631        Never-married        Tech-support   Not-in-family   
13084        Never-married     Priv-house-serv   Not-in-family   
15059        Never-married     Farming-fishing   Not-in-family   
17040   Married-civ-spouse        Craft-repair         Husband   
18555        Never-married       Other-service   Not-in-family   
18698        Never-married     Farming-fishing   Not-in-family   
21318        Never-married        Adm-clerical       Own-child   
21490        Never-married     Exec-managerial       Own-child   
21875   Married-civ-spouse        Craft-repair         Husband   
22300        Never-married     Priv-house-serv   Not-in-family   
22367        Never-married      Prof-specialty   Not-in-family   
22494   Married-civ-spouse        Craft-repair         Husband   
25872        Never-married   Handlers-cleaners   Not-in-family   
26313        Never-married      Prof-specialty   Not-in-family   
28230        Never-married   Machine-op-inspct   Not-in-family   
28522   Married-civ-spouse      Prof-specialty         Husband   
28846   Married-civ-spouse        Craft-repair         Husband   
29157   Married-civ-spouse   Machine-op-inspct         Husband   
30845             Divorced        Adm-clerical       Unmarried   
31993        Never-married       Other-service       Own-child   
32404             Divorced       Other-service   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
4881                 White     Male             0             0   
5104    Asian-Pac-Islander     Male             0             0   
9171                 White   Female             0             0   
11631                White   Female             0             0   
13084                White   Female             0             0   
15059                White     Male             0             0   
17040                White     Male             0             0   
18555                Black     Male         

In [211]:
df2 = df_dup.query('Sex == Male')

UndefinedVariableError: name 'Sex' is not defined